File: basic/go/sync.go
package main

import (
	"encoding/json"
	"io/ioutil"
	"log"
	"net/http"

	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
)

type SyncRequest struct {
	Parent   unstructured.Unstructured `json:"parent"`
	Children Children                  `json:"children"`
}

type Children struct {
	Pongs []unstructured.Unstructured `json:"example.com/v1.Pong"`
}

type SyncResponse struct {
	Children []unstructured.Unstructured `json:"children"`
}

func sync(req *SyncRequest) (*SyncResponse, error) {
	res := &SyncResponse{}
	name, found, err := unstructured.NestedString(req.Parent.UnstructuredContent(),
		"spec", "name")
	if err != nil || !found {
		name = "Unknown"
	}

	chield := unstructured.Unstructured{
		Object: map[string]interface{}{
			"apiVersion": "example.com/v1",
			"kind":       "Pong",
			"metadata": map[string]interface{}{
				"name": req.Parent.GetName(),
			},
			"spec": map[string]interface{}{
				"message": "Hello " + name,
			},
		},
	}
	res.Children = append(res.Children, chield)
	return res, nil
}

func syncHandler(w http.ResponseWriter, r *http.Request) {
	body, err := ioutil.ReadAll(r.Body)
	if err != nil {
		http.Error(w, err.Error(), http.StatusInternalServerError)
		return
	}
	req := &SyncRequest{}
	if err := json.Unmarshal(body, req); err != nil {
		http.Error(w, err.Error(), http.StatusBadRequest)
		return
	}
	res, err := sync(req)
	if err != nil {
		http.Error(w, err.Error(), http.StatusInternalServerError)
		return
	}
	body, err = json.Marshal(&res)
	if err != nil {
		http.Error(w, err.Error(), http.StatusInternalServerError)
		return
	}
	w.Header().Set("Content-Type", "application/json")
	w.Write(body)
}

func main() {
	http.HandleFunc("/sync", syncHandler)

	log.Fatal(http.ListenAndServe(":8080", nil))
}

File: basic/js/sync.js
const express = require("express");
const bodyParser = require('body-parser');

const PORT = 8080;
const app = express();
app.use(bodyParser.urlencoded({ extended: false }));
app.use(bodyParser.json())

app.post("/sync", (req, res) => {
    let observed = req.body;
    let desired = {children: []};

    try {
        // Observed ping object
        let ping = observed.parent;
        let name = "Unknown"
        try{
            name = ping.spec.name
        }catch(e){
            console.log(e.stack)
        }

        let pong =  {
            apiVersion: 'example.com/v1',
            kind: 'Pong',
            metadata: {
                name: ping.metadata.name
            },
            spec: {
                message: "Hello "+ name +" !!"
            }
        }

        // Generate desired children
        desired.children = [
            pong
        ];

        res.status(200).send(desired)
    } catch (e) {
        console.log(e.stack)
        res.status(500).send({body: e.stack})
    }
});

app.listen(PORT, () => {
    console.log(`Server is listening on port: ${PORT}`);
});
